solidity/contracts/CheckpointFraudProofs.sol

```diff
@@ -138,4 +145,29 @@ contract CheckpointFraudProofs {
         );
         return root != checkpoint.root;
     }
+
+    function markPremature(Checkpoint calldata checkpoint) external {
+        require(isPremature(checkpoint), "must be premature");
+        _markFraudulent(checkpoint);
+    }
```

Comment:
@ltyu and I discussed an edge case where an attacker successfully 
1. predicts checkpoints from message dispatches pending in the mempool
2. frontruns them with `markPremature` calls
3. message dispatches land on the origin chain
4. valid checkpoints are considered fraudulent in this contract's storage mapping

Reply:
I think the only mitigation is removing this fraud caching layer for the prematurity condition -- we must atomically check this condition when we are slashing